From: Jiri Slaby <jirislaby@gmail.com>
Date: Fri, 14 Aug 2026 08:06:26 +0200
Subject: symbols: use non-debug BFD to retrieve .rodata
References: https://github.com/crash-utility/crash/issues/232
Git-repo: https://github.com/crash-utility/crash
Git-commit: abcb45e7d443e693a59a759ca2862530e917c891
Patch-mainline: yes

When running crash on openSUSE, get_linux_banner_from_vmlinux() returns
success but fill the target buffer with all zero bytes (`\0`).

This happens because `st->bfd` is initially opened for the main binary
(`vmlinux`), but is later overwritten with the debuginfo file
(`vmlinux.debug`) in `check_gnu_debuglink()`. In separate debug files,
`.rodata` has no content, so is marked only as `ALLOC` without
`SEC_HAS_CONTENTS`.

When `bfd_get_section_contents()` is called on a section without
`SEC_HAS_CONTENTS`, BFD clears the buffer to zeros and returns TRUE. As
a result, `get_linux_banner_from_vmlinux()` thinks it successfully read
the banner, while it actually received zeroed bytes.

Fix this by caching the original main executable's BFD (into
`st->bfd_orig`) before `st->bfd` gets swapped for the debuginfo file.

`get_linux_banner_from_vmlinux()` will then fall back to `st->bfd_orig`
if present, ensuring `.rodata` contents are read from the binary that
actually contains the raw section data.

Fixes #232.

Signed-off-by: Jiri Slaby <jirislaby@gmail.com>
Cc: Dave Young <yangrr.2009@tsinghua.org.cn>
Cc: <ltao@redhat.com>
Cc: <mpilaniy@redhat.com>
Reviewed-by: Dave Young <yangrr.2009@tsinghua.org.cn>
Signed-off-by: Dave Young <ruirui.yang@linux.dev>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 defs.h    | 1 +
 kernel.c  | 5 +++--
 symbols.c | 1 +
 3 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/defs.h b/defs.h
index ef34db90113f..134e7597e966 100644
--- a/defs.h
+++ b/defs.h
@@ -2924,6 +2924,7 @@ struct symbol_table_data {
 #ifdef GDB_5_3
 	struct _bfd *bfd;
 #else
+	struct bfd *bfd_orig;
 	struct bfd *bfd;
 #endif
 	struct sec *sections;
diff --git a/kernel.c b/kernel.c
index cf5e5bcfe785..723b88e90cb4 100644
--- a/kernel.c
+++ b/kernel.c
@@ -12164,6 +12164,7 @@ check_vmcoreinfo(void)
 static
 int get_linux_banner_from_vmlinux(char *buf, size_t size)
 {
+	struct bfd *bfd = st->bfd_orig ? : st->bfd;
 	struct bfd_section *sect;
 	long offset;
 	ulong start_rodata;
@@ -12175,7 +12176,7 @@ int get_linux_banner_from_vmlinux(char *buf, size_t size)
 	else
 		return FALSE;
 
-	sect = bfd_get_section_by_name(st->bfd, ".rodata");
+	sect = bfd_get_section_by_name(bfd, ".rodata");
 	if (!sect)
 		return FALSE;
 
@@ -12187,7 +12188,7 @@ int get_linux_banner_from_vmlinux(char *buf, size_t size)
 	 */
 	offset = symbol_value("linux_banner") - start_rodata;
 
-	if (!bfd_get_section_contents(st->bfd,
+	if (!bfd_get_section_contents(bfd,
 				      sect,
 				      buf,
 				      offset,
diff --git a/symbols.c b/symbols.c
index 1515385d5724..270e14e26d28 100644
--- a/symbols.c
+++ b/symbols.c
@@ -444,6 +444,7 @@ check_gnu_debuglink(bfd *bfd)
 	return FALSE;
 
 reset_bfd:
+	st->bfd_orig = st->bfd;
 
         if ((st->bfd = bfd_openr(pc->debuginfo_file, NULL)) == NULL)
                 error(FATAL, "cannot open object file: %s\n", 
-- 
2.55.0

